src/txmempool.cpp

```diff
@@ -266,6 +266,7 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry,
                                            std::string &errString,
                                            bool fSearchForParents /* = true */) const
 {
+    setAncestors.clear();
```

Comment:
can you explain why this is a bugfix? There is currently no code path that doesn't clear the set before calling this function.

Reply:
Ok, I see it is only in the package acceptance code path?

Reply:
> it is only in the package acceptance code path?

That's my understanding, yes. `MemPoolAccept::PreChecks` and `MemPoolAccept::SubmitPackage` are the only functions where setAncestors isn't (always) cleared before calling `CalculateMemPoolAncestors`. In `MemPoolAccept::AcceptMultipleTransactions()`, we first call `PreChecks()` and then `SubmitPackage()` with the same `Workspace`. `PreChecks` always seems to be called with a fresh `Workspace`, so I'd say `SubmitPackage` is the only issue. Given that this is currently only used in `testmempoolaccept` and `submitpackage` RPCs, it shouldn't be a particularly impactful bug.

I'll update the PR description a bit more now, and if I have to retouch, I'll update the bugfix commit message to add some more background.

Reply:
Correct, the only path is in `SubmitPackage`. It's not problematic to pass in `ws.m_ancestors` that is already populated, but I agree CMPA *should* always start with an empty ancestor set. Feel free to call it a bug or unexpected behavior etc.

Reply:
I wrote a quick check on `master` to verify that ancestors passed into CMPA are indeed always a subset of what is returned at the end of CMPA. At least as far as the scenarios covered in the unit and functional tests, that seems to be the case, since with this patch all tests complete without throwing.

<details>
<summary>git diff on master (5974c49f9)</summary>

```diff
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index 84ed2e9ef..e024ebe81 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -190,6 +190,8 @@ bool CTxMemPool::CalculateAncestorsAndCheckLimits(size_t entry_size,
                                                   const Limits& limits,
                                                   std::string &errString) const
 {
+    setEntries incoming_ancestors{setAncestors.begin(), setAncestors.end()};
+    setAncestors.clear();
     size_t totalSizeWithAncestors = entry_size;
 
     while (!staged_ancestors.empty()) {
@@ -226,6 +228,9 @@ bool CTxMemPool::CalculateAncestorsAndCheckLimits(size_t entry_size,
         }
     }
 
+    for (auto& ancestor : incoming_ancestors){
+        if (setAncestors.find(ancestor) == setAncestors.end()) throw std::runtime_error("incoming ancestor not found");
+    }
     return true;
 }
 ```
</details>

Reply:
Why is this bugfix added here and then removed later?

Reply:
I'm unsure if we should keep the bugfix commit. As glozow says, the interface was brittle and could have been used in a buggy way, but that currently doesn't appear to be the case (as per my own verification too). With that understanding, we probably don't want to backport this? Alternatively, the separate bugfix commit could make it easier for downstream projects to patch this (potential) issue without having to adopt the full changeset? But I'm not sure that's how we usually do things.

cc @luke-jr - any further thoughts on this after [your earlier comment](https://github.com/bitcoin/bitcoin/pull/26289#issuecomment-1290725923) and [glozow's explanation](https://github.com/bitcoin/bitcoin/pull/26289#discussion_r1007899547)?